docs/tutorials/demo_best_practices.py

```diff
@@ -0,0 +1,281 @@
+    "weather-2010.json",
+)
+urllib.request.urlretrieve(
+    "https://raw.githubusercontent.com/aai-institute/lakefs-spec/main/docs/tutorials/.lakectl.yaml",
+    os.path.expanduser("~/.lakectl.yaml"),
+)
+# %% [markdown]
+"""
+## Define a data repository
+
+We got the data for the year 2010. That should be enough for initial prototyping.
+Later, however, we want to use more data. Since our dataset will be evolving, we implement data version control.
+This ensure the reproducibility of our experiments, enables collaboration with colleagues, and ensures our dynamic dataset to stay a valuable asset.
+
+To set up our versioning, we need to decide on a versioning tool, set up a repository, and define which data is considered in scope and should be versioned and which is not.
+
+Let us now set up a lakeFS repository using lakeFS-spec, our filesystem implementation for lakeFS.
+"""
+
+# %%
+fs = lakefs_spec.LakeFSFileSystem()
+
+REPO_NAME = "weatherpred"
+
+repo = lakefs.Repository(REPO_NAME, fs.client).create(storage_namespace=f"local://{REPO_NAME}")
+
+
+# %% [markdown]
+"""
+lakeFS works similarly to `git` as a versioning system.
+You can create *commits* that contain specific changes to the data.
+You can also work with *branches* to create an isolated view of the data.
+Every commit (on any branch) is identified by a commit SHA, a unique identifier obtained via a hashing function.
+
+## Branching Strategy
+
+We recommend following a branching strategy that ensures the data integrity on the main branch.
+Since we are about to do some data wrangling, we will branch off `main` and later merge back into it, once we are sure everything works as expected.
+"""
+
+# %%
+NEW_BRANCH_NAME = "transform-raw-data"
+
+with fs.transaction as tx:
+    fs.put(outfile, f"{REPO_NAME}/{NEW_BRANCH_NAME}/weather-2010.json")
+    tx.commit(repository=REPO_NAME, branch=NEW_BRANCH_NAME, message="Add 2010 weather data")
+
+# %% [markdown]
+"""
+## Utilize reusable and tested functions for data transformation
+
+Now that we have the data on the `transform-raw-data` branch, we can start with the transformation.
+
+It is good practice to encapsulate common transformations in functions.
+"""
+
+
+# %%
+def load_json_data(filepath):
+    if hasattr(filepath, "close") and hasattr(filepath, "tell"):
+        return json.load(filepath)
+    else:
+        with open(filepath, "r") as f:
+            return json.load(f)
+
+
+# %%
+def create_dataframe_from_json(json_data):
+    df = pd.DataFrame.from_dict(json_data["hourly"])
+    return df
+
+
+sample_json_data = {
+    "hourly": [{"time": "2023-01-01 00:00", "rain": 0}, {"time": "2023-01-01 01:00", "rain": 2}]
+}
+
+# Tests
```

Comment:
I wonder whether the tests should be split off into separate code cells for visual distinction